### PR TITLE
WIP: Add Asset to Vega Spec

### LINKED ISSUE #225: Adding assets on Vega
In GitLab by @tamlyn10 on Mar 20, 2020, 12:56

Write a spec for adding assets on Vega.

https://vegaprotocol.slack.com/archives/CAHA5EX0F/p1584706665085200

### PR DESCRIPTION
**Pull request by C0deMunk33**
Add asset to Vega from Ethereum bridges

### CODE REVIEW

**Comment by C0deMunk33** [comment]
I need some input on this one. I got the Ethereum side of this pretty well understood, but it will need info about the Vega side. Thanks!

**Comment by C0deMunk33** [comment]
For more info, see issue #225 and the contained thread: 
https://github.com/vegaprotocol/product/issues/225

File: specs/0000-add-asset-to-vega.md [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+Start date: 2020-04-09
+Specification PR: https://gitlab.com/vega-protocol/product/merge_requests
+
+# Summary
+In order for the Vega network to hold value via asset bridges, assets must be added to Vega and that order must be propagated to the appropriate Vega bridge smart contract.  
+
+# Guide-level explanation
+### Ethereum-based assets
+#### ERC20
+To add an ERC20 token-based asset to Vega, a market maker or other interested party will submit the target ethereum token address to the Vega interface (`TODO`).
```
Comment (barnabee) [inline specs/0000-add-asset-to-vega.md]: They will also need to submit the other details (names, short codes, whatever) that need to be defined for the asset. In general I don't know if this spec can be approved without a spec for the asset framework?

Comment (barnabee) [reply specs/0000-add-asset-to-vega.md]: The Vega transaction also needs to be defined and how an asset gets approved on Vega (statuses, etc.)

File: specs/0000-add-asset-to-vega.md [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+Once submitted, the request is put through Vega consensus. 
+Once accepted through consensus, a multisig signature bundle is aggregated from Vega validator nodes and the signature bundle is then submitted to the appropriate Vega ERC20 bridge by way of the `whitelist_asset` function.
```
Comment (barnabee) [inline specs/0000-add-asset-to-vega.md]: I think the asset will need to be *voted* in through a governance action not jsut accepted through consensus in order to be whitelisted?

File: specs/0000-add-asset-to-vega.md [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+Upon successful execution of the `whitelist_asset` function, that token will be available for on-chain deposits via the `deposit_asset` function on the smart contract.
+Deposits that are made to the contract will raise the `Asset_Deposited` event which will then be consumed and propagated through Vega consensus by way of the Event Queue.
+ 
+#### ERC1155
```
Comment (barnabee) [inline specs/0000-add-asset-to-vega.md]: Same comments as above

File: specs/0000-add-asset-to-vega.md [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+To add an ERC1155 token-based asset to Vega, a market maker or other interested party will submit the target ethereum smart contract address and asset ID (uint256) to the Vega interface (`TODO`).
+Once submitted, the request is put through Vega consensus. 
+Once accepted through consensus, a multisig signature bundle is aggregated from Vega validator nodes and the signature bundle is then submitted to the appropriate Vega ERC1155 bridge by way of the `whitelist_asset` function.
+Upon successful execution of the `whitelist_asset` function, that token will be available for on-chain deposits via the `deposit_asset` function on the smart contract.
+Deposits that are made to the contract will raise the `Asset_Deposited` event which will then be consumed and propagated through Vega consensus by way of the Event Queue.
+
+### Other blockchains
+`TODO`
+
+# Reference-level explanation
```
Comment (barnabee) [inline specs/0000-add-asset-to-vega.md]: This needs to be finished and describe the workflow/state machine withion Vega for proposing and approving adding assets as well as how it interacts with the bridge contract.

**Review by barnabee — DISMISSED** [review]
Unfinished spec needs to be finished, some comments to address too. We also need a spec for the asset framework (fields stored and proposed/voted on for an asset) OR to include that in this spec (I would then rename the spec asset framework though)

File: specs/0000-add-asset-to-vega.md (hunk 1 above)
Comment (C0deMunk33) [reply specs/0000-add-asset-to-vega.md]: This spec is getting changed to Asset Framework based on your comments. I'll have updates soon(tm)

**Comment by C0deMunk33** [comment]
I've updated this quite a bit. There's TODOs in most of the places I can't do (and a few I haven't gotten to yet).

Most of the unknowns come from the interface to both submitting things to the Vega API (like withdrawal requests) or getting signature aggregation, which I don't think is fully fleshed out. Both of those are things I will probably get from @jeremyletang

File: specs/0000-add-asset-to-vega.md (hunk 2 above)
Comment (C0deMunk33) [reply specs/0000-add-asset-to-vega.md]: Indeed, I expanded this out substantially to explain how governance does what it do (with a few TODOs)

File: specs/0000-add-asset-to-vega.md (hunk 4 above)
Comment (C0deMunk33) [reply specs/0000-add-asset-to-vega.md]: Indeed, added these sections. Expanding them now.

File: specs/0000-add-asset-to-vega.md (hunk 1 above)
Comment (C0deMunk33) [reply specs/0000-add-asset-to-vega.md]: I've expanded this out with the current list of parameters for an asset. More todo, but it's coming along.

As for statuses, had a chat and status is currently either "listed" or "not listed" and this is implicit by it's existence on-chain.

@jeremyletang  what about a scenario that the asset has been delisted and we still need to process withdrawals for remaining balances.?

**Review by barnabee — COMMENTED** [review]
Looks a lot better this way I think. Will review in more detail once it's closer to done but no obvious issues jumping out with what's there.

**Comment by C0deMunk33** [comment]
> Looks a lot better this way I think. Will review in more detail once it's closer to done but no obvious issues jumping out with what's there.

Awesome, thanks @barnabee  There's a couple TODOs left, but I think this works.

**Comment by C0deMunk33** [comment]
@jeremyletang at your convenience please look at the TODOs in the spec.

**Review by C0deMunk33 — COMMENTED** [review]
Looking good

**Comment by tamlyn10** [comment]
@jeremyletang will review and merge or close

**Comment by tamlyn10** [comment]
@barnabee to do one final review. @jeremyletang to finish remaining TODOs and merge.

**Comment by tamlyn10** [comment]
as above @jeremyletang

**Comment by tamlyn10** [comment]
Review Jeremy's changes @barnabee @C0deMunk33

**Comment by C0deMunk33** [comment]
Looks good to me, I say ready to merge.

**Pull request merged**